scene/resources/curve.cpp

```diff
@@ -1630,114 +1698,150 @@ Vector3 Curve3D::sample_baked(real_t p_offset, bool p_cubic) const {
+	Curve3D::Interval interval = _find_interval(p_offset);
+	return _sample_baked(interval, p_cubic);
+}
+
+Transform3D Curve3D::sample_baked_with_rotation(real_t p_offset, bool p_cubic, bool p_apply_tilt) const {
+	if (baked_cache_dirty) {
+		_bake();
 	}
 
-	if (idx == count - 1) {
-		return p_apply_tilt ? r[idx].rotated((rp[idx] - rp[idx - 1]).normalized(), rt[idx]) : r[idx];
+	// Validate: Curve may not have baked points.
+	const int point_count = baked_point_cache.size();
+	ERR_FAIL_COND_V_MSG(point_count == 0, Transform3D(), "No points in Curve3D.");
+
+	if (point_count == 1) {
+		Transform3D t;
+		t.origin = baked_point_cache.get(0);
+		ERR_FAIL_V_MSG(t, "Only 1 point in Curve3D.");
 	}
 
-	real_t offset_begin = baked_dist_cache[idx];
-	real_t offset_end = baked_dist_cache[idx + 1];
+	p_offset = CLAMP(p_offset, 0.0, get_baked_length()); // PathFollower implement wrapping logic.
 
-	real_t idx_interval = offset_end - offset_begin;
-	ERR_FAIL_COND_V_MSG(p_offset < offset_begin || p_offset > offset_end, Vector3(0, 1, 0), "Couldn't find baked segment.");
+	// 0. Find interval for all sampling steps.
+	Curve3D::Interval interval = _find_interval(p_offset);
 
-	real_t frac = (p_offset - offset_begin) / idx_interval;
+	// 1. Sample position.
+	Vector3 pos = _sample_baked(interval, p_cubic);
+
+	// 2. Sample rotation frame.
+	Basis frame = _sample_posture(interval, p_apply_tilt);
+
+	return Transform3D(frame, pos);
+}
+
+real_t Curve3D::sample_baked_tilt(real_t p_offset) const {
+	if (baked_cache_dirty) {
+		_bake();
+	}
 
-	Vector3 forward = (rp[idx + 1] - rp[idx]).normalized();
-	Vector3 up = r[idx];
-	Vector3 up1 = r[idx + 1];
+	// Validate: Curve may not have baked tilts.
+	int pc = baked_tilt_cache.size();
+	ERR_FAIL_COND_V_MSG(pc == 0, 0, "No tilts in Curve3D.");
 
-	if (p_apply_tilt) {
-		up.rotate(forward, rt[idx]);
-		up1.rotate(idx + 2 >= count ? forward : (rp[idx + 2] - rp[idx + 1]).normalized(), rt[idx + 1]);
+	if (pc == 1) {
+		return baked_tilt_cache.get(0);
 	}
 
-	Vector3 axis = up.cross(up1);
+	p_offset = CLAMP(p_offset, 0.0, get_baked_length()); // PathFollower implement wrapping logic
 
-	if (axis.length_squared() < CMP_EPSILON2) {
-		axis = forward;
-	} else {
-		axis.normalize();
+	Curve3D::Interval interval = _find_interval(p_offset);
+	return _sample_baked_tilt(interval);
+}
+
+Vector3 Curve3D::sample_baked_up_vector(real_t p_offset, bool p_apply_tilt) const {
+	if (baked_cache_dirty) {
+		_bake();
+	}
+
+	// Validate: Curve may not have baked up vectors.
+	ERR_FAIL_COND_V_MSG(!up_vector_enabled, Vector3(0, 1, 0), "No up vectors in Curve3D.");
+
+	int count = baked_up_vector_cache.size();
+	if (count == 1) {
+		return baked_up_vector_cache.get(0);
 	}
 
-	return up.rotated(axis, up.angle_to(up1) * frac);
+	p_offset = CLAMP(p_offset, 0.0, get_baked_length()); // PathFollower implement wrapping logic
+
+	Curve3D::Interval interval = _find_interval(p_offset);
+	return _sample_posture(interval, p_apply_tilt).get_column(1);
```

Comment:
This return is strange given the name of the function `sample_baked_up_vector()`. This affects other classes that use Curve.

If you need an up_vector modified by `_sample_posture()`, you should use another function instead of `sample_baked_up_vector()`.

Reply:
`_sample_posture()` doesn't modify posture, it just interpolate the "pure" baked posture . The modification is done on Pathfollow3D

The reason of calling `_sample_posture()` `sample_baked_up_vector()` are
1. keep one implementation for consistency
2. the best of practice of interpolate rotation vectors is not linear interpolation, but through basis or equation

Reply:
This is a redundant process when only up_vector is needed. If it is really necessary, the baked_up_vector_chace itself must be pre-modified in _bake().

Reply:
> It is a redundant process when only up_vector is needed

No, it is not redundant. For rotation vector(like up vector) simple linear interpolation produce `inaccurate` result. Even if you don't mind inaccuracy, it will be `inconsistent` with the up vector you get from `sample_baked_with_rotation`, which use basis interpolation(more accurate method).

> If it is really necessary, the baked_up_vector_chace itself must be pre-modified in _bake().

`_sample_posture()` do not modify posture, the up vector produced from it, is conceptually equivalent with direct linear interpolation, just more accurate.

Reply:
If you really need this, you should rename this function from `sample_baked_up_vector()` to `sample_up_vector()`. Then we can determine if it is acceptable by others.

Reply:
Although implementation is changed, `sample_baked_up_vector` is still an appropriate name. After all the the up vector is still calculated from `baked_up_vector_cache`, just with different method.

Reply:
Since the up_vectors don't have rotation, I'm not sure how much sense it makes to consider a forward_vector for something other than a PathFollow3D. It would increase the computational cost. But well... maybe it is acceptable if the behavior is fine.